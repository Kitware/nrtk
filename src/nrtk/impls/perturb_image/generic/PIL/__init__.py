"""Module for PIL based implementations of PerturbImage"""
