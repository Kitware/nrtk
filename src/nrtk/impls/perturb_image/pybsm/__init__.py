"""Module for PyBSM based implementations of PerturbImage"""
