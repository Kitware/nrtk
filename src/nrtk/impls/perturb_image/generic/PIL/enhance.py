from typing import TYPE_CHECKING, Any, Dict, Optional, Protocol, Type, runtime_checkable

import numpy as np
from PIL import Image, ImageEnhance
from PIL.Image import Image as PILImage

from nrtk.interfaces.perturb_image import PerturbImage


@runtime_checkable
class _Enhancement(Protocol):  # Used for type checking only  # pragma: no cover
    def __init__(self: "_Enhancement", image: PILImage):
        pass

    def enhance(self: "_Enhancement", factor: float) -> PILImage:
        pass


class _PILEnhancePerturber(PerturbImage):
    def __init__(self, factor: float = 1.0):
        """:param factor: Enhancement factor."""
        if factor < 0.0:
            raise ValueError(f"{type(self).__name__} invalid factor ({factor})." f" Must be >= 0.0")

        self.factor = factor

    def _perturb(
        self,
        enhancement: Type[_Enhancement],
        image: np.ndarray,
        additional_params: Optional[Dict[str, Any]] = None,
    ) -> np.ndarray:
        """Call appropriate enhancement interface and perform any necessary data type conversion.

        :param enhancement: Ehancement to apply.
        :param image: Input image as a numpy array.

        :return: Peturbed image as numpy array, including matching shape and dtype.
        """
        if additional_params is None:
            additional_params = dict()
        dtype = image.dtype
        # PIL does not support RGB floating point images so we must do an
        # intermediary conversion
        if np.issubdtype(dtype, np.floating):
            image = (image * 255).astype(np.uint8)

        image_pil = Image.fromarray(image)
        image_enhanced = enhancement(image_pil).enhance(factor=self.factor)
        image_np = np.array(image_enhanced)

        # Convert back to floating point dtype if needed
        if np.issubdtype(dtype, np.floating):
            image_np = image.astype(dtype) / 255

        return image_np

    def get_config(self) -> Dict[str, Any]:
        return {"factor": self.factor}


class BrightnessPerturber(_PILEnhancePerturber):
    """Adjusts image stimulus brightness."""

    def perturb(self, image: np.ndarray, additional_params: Optional[Dict[str, Any]] = None) -> np.ndarray:
        """Return image stimulus with adjusted brightness."""
        if additional_params is None:
            additional_params = dict()
        enhancement = ImageEnhance.Brightness
        if TYPE_CHECKING and not isinstance(enhancement, _Enhancement):  # pragma: no cover
            raise ValueError("enhancement does not conform to _Enhancement protocol")
        return self._perturb(enhancement=enhancement, image=image)


class ColorPerturber(_PILEnhancePerturber):
    """Adjusts image stimulus color balance."""

    def perturb(self, image: np.ndarray, additional_params: Optional[Dict[str, Any]] = None) -> np.ndarray:
        """Return image stimulus with adjusted color balance."""
        if additional_params is None:
            additional_params = dict()
        enhancement = ImageEnhance.Color
        if TYPE_CHECKING and not isinstance(enhancement, _Enhancement):  # pragma: no cover
            raise ValueError("enhancement does not conform to _Enhancement protocol")
        return self._perturb(enhancement=enhancement, image=image)


class ContrastPerturber(_PILEnhancePerturber):
    """Adjusts image stimulus contrast."""

    def perturb(self, image: np.ndarray, additional_params: Optional[Dict[str, Any]] = None) -> np.ndarray:
        """Return image stimulus with adjusted contrast."""
        if additional_params is None:
            additional_params = dict()
        enhancement = ImageEnhance.Contrast
        if TYPE_CHECKING and not isinstance(enhancement, _Enhancement):  # pragma: no cover
            raise ValueError("enhancement does not conform to _Enhancement protocol")
        return self._perturb(enhancement=enhancement, image=image)


class SharpnessPerturber(_PILEnhancePerturber):
    """Adjusts image stimulus sharpness."""

    def __init__(self, factor: float = 1.0):
        """:param rng: Enhancement factor."""
        if factor < 0.0 or factor > 2.0:
            raise ValueError(f"{type(self).__name__} invalid sharpness factor ({factor})." f" Must be in [0.0, 2.0]")

        super().__init__(factor=factor)

    def perturb(self, image: np.ndarray, additional_params: Optional[Dict[str, Any]] = None) -> np.ndarray:
        """Return image stimulus with adjusted sharpness."""
        if additional_params is None:
            additional_params = dict()
        enhancement = ImageEnhance.Sharpness
        if TYPE_CHECKING and not isinstance(enhancement, _Enhancement):  # pragma: no cover
            raise ValueError("enhancement does not conform to _Enhancement protocol")
        return self._perturb(enhancement=enhancement, image=image)
