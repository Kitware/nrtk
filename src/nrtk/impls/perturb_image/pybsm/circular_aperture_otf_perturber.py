from __future__ import annotations

from typing import Any, Dict, Optional, Sequence, Type, TypeVar

try:
    import cv2

    cv2_available = True
except ImportError:
    cv2_available = False
import numpy as np
import pybsm.radiance as radiance
from pybsm.otf.functional import (
    circular_aperture_OTF,
    otf_to_psf,
    resample_2D,
    weighted_by_wavelength,
)
from pybsm.utils import load_database_atmosphere, load_database_atmosphere_no_interp
from smqtk_core.configuration import (
    from_config_dict,
    make_default_config,
    to_config_dict,
)

from nrtk.impls.perturb_image.pybsm.scenario import PybsmScenario
from nrtk.impls.perturb_image.pybsm.sensor import PybsmSensor
from nrtk.interfaces.perturb_image import PerturbImage

C = TypeVar("C", bound="CircularApertureOTFPerturber")


class CircularApertureOTFPerturber(PerturbImage):
    def __init__(
        self,
        sensor: Optional[PybsmSensor] = None,
        scenario: Optional[PybsmScenario] = None,
        mtf_wavelengths: Optional[Sequence[float]] = None,
        mtf_weights: Optional[Sequence[float]] = None,
        interp: Optional[bool] = True,
    ) -> None:
        """Initializes the CircularApertureOTFPerturber.

        :param name: string representation of object
        :param sensor: pyBSM sensor object
        :param scenario: pyBSM scenario object
        :param mtf_wavelengths: a numpy array of wavelengths (m)
        :param mtf_wavelengths_weights: a numpy array of weights for each wavelength contribution (arb)
        :param interp: a boolean determinings whether load_database_atmosphere is used with or without
                       interpoloation

        If both sensor and scenario parameters are absent, then default values
        will be used for their parameters

        If none of mtf_wavelengths, mtf_weights, sensor or scenario parameters are provided, the values
        of mtf_wavelengths and mtf_weights will default to [0.50e-6, 0.66e-6] and [1.0, 1.0] respectively

        If sensor and scenario parameters are provided, but not mtf_wavelengths and mtf_weights, the
        values of mtf_wavelengths and mtf_weights will come from the sensor and scenario objects.

        If mtf_wavelengths and mtf_weights are provided by the user, those values will be used
        in the otf caluclattion

        :raises: ValueError if mtf_wavelengths and mtf_weights are not equal length
        :raises: ValueError if mtf_wavelengths is empty or mtf_weights is empty
        """
        if not self.is_usable():
            raise ImportError("OpenCV not found. Please install 'nrtk[graphics]' or 'nrtk[headless]'.")

        if sensor and scenario:
            if interp:
                atm = load_database_atmosphere(scenario.altitude, scenario.ground_range, scenario.ihaze)
            else:
                atm = load_database_atmosphere_no_interp(scenario.altitude, scenario.ground_range, scenario.ihaze)
            (
                _,
                _,
                spectral_weights,
            ) = radiance.reflectance_to_photoelectrons(atm, sensor.create_sensor(), sensor.int_time)

            wavelengths = spectral_weights[0]
            weights = spectral_weights[1]

            # cut down the wavelength range to only the regions of interests
            pos_weights = np.where(weights > 0.0)
            self.mtf_wavelengths = (
                np.asarray(mtf_wavelengths) if mtf_wavelengths is not None else wavelengths[pos_weights]
            )
            self.mtf_weights = np.asarray(mtf_weights) if mtf_weights is not None else weights[pos_weights]

            D = sensor.D  # noqa: N806
            eta = sensor.eta  # noqa: N806

            self.slant_range = np.sqrt(scenario.altitude**2 + scenario.ground_range**2)
            self.ifov = (sensor.p_x + sensor.p_y) / 2 / sensor.f
        else:
            self.mtf_wavelengths = (
                np.asarray(mtf_wavelengths)
                if mtf_wavelengths is not None
                else np.array([0.58 - 0.08, 0.58 + 0.08]) * 1.0e-6
            )
            self.mtf_weights = (
                np.asarray(mtf_weights) if mtf_weights is not None else np.ones(len(self.mtf_wavelengths))
            )
            # Assume visible spectrum of light
            self.ifov = -1
            self.slant_range = -1
            # Default value for lens diameter and relative linear obscuration
            D = 0.003  # noqa: N806
            eta = 0.0  # noqa: N806

        if self.mtf_wavelengths.size == 0:
            raise ValueError("mtf_wavelengths is empty")

        if self.mtf_weights.size == 0:
            raise ValueError("mtf_weights is empty")

        if self.mtf_wavelengths.size != self.mtf_weights.size:
            raise ValueError("mtf_wavelengths and mtf_weights are not the same length")

        self.sensor = sensor
        self.scenario = scenario
        self.interp = interp

        # Assume if nothing else cuts us off first, diffraction will set the
        # limit for spatial frequency that the imaging system is able
        # to resolve is (1/rad).
        cutoff_frequency = D / np.min(self.mtf_wavelengths)
        u_rng = np.linspace(-1.0, 1.0, 1501) * cutoff_frequency
        v_rng = np.linspace(1.0, -1.0, 1501) * cutoff_frequency

        # meshgrid of spatial frequencies out to the optics cutoff
        uu, vv = np.meshgrid(u_rng, v_rng)
        self.df = (abs(u_rng[1] - u_rng[0]) + abs(v_rng[0] - v_rng[1])) / 2

        def ap_function(wavelengths: float) -> np.ndarray:
            return circular_aperture_OTF(uu, vv, wavelengths, D, eta)  # noqa: E731

        self.ap_OTF = weighted_by_wavelength(self.mtf_wavelengths, self.mtf_weights, ap_function)

    def perturb(self, image: np.ndarray, additional_params: Optional[Dict[str, Any]] = None) -> np.ndarray:
        """:raises: ValueError if 'img_gsd' not present in additional_params"""
        if additional_params is None:
            additional_params = dict()
        if self.ifov >= 0 and self.slant_range >= 0:
            if "img_gsd" not in additional_params:
                raise ValueError(
                    "'img_gsd' must be present in image metadata\
                                  for this perturber"
                )
            ref_gsd = additional_params["img_gsd"]
            psf = otf_to_psf(self.ap_OTF, self.df, 2 * np.arctan(ref_gsd / 2 / self.slant_range))

            # filter the image
            blur_img = cv2.filter2D(image, -1, psf)

            # resample the image to the camera's ifov
            sim_img = resample_2D(blur_img, ref_gsd / self.slant_range, self.ifov)

        else:
            # Default is to set dxout param to same value as dxin
            psf = otf_to_psf(self.ap_OTF, self.df, 1 / (self.ap_OTF.shape[0] * self.df))

            sim_img = cv2.filter2D(image, -1, psf)

        return sim_img.astype(np.uint8)

    def __call__(self, image: np.ndarray, additional_params: Optional[Dict[str, Any]] = None) -> np.ndarray:
        """Alias for :meth:`.NIIRS.apply`."""
        if additional_params is None:
            additional_params = dict()
        return self.perturb(image, additional_params)

    @classmethod
    def get_default_config(cls) -> Dict[str, Any]:
        cfg = super().get_default_config()
        cfg["sensor"] = make_default_config([PybsmSensor])
        cfg["scenario"] = make_default_config([PybsmScenario])
        return cfg

    @classmethod
    def from_config(cls: Type[C], config_dict: Dict, merge_default: bool = True) -> C:
        config_dict = dict(config_dict)
        sensor = config_dict.get("sensor", None)
        if sensor is not None:
            config_dict["sensor"] = from_config_dict(sensor, [PybsmSensor])
        scenario = config_dict.get("scenario", None)
        if scenario is not None:
            config_dict["scenario"] = from_config_dict(scenario, [PybsmScenario])

        return super().from_config(config_dict, merge_default=merge_default)

    def get_config(self) -> Dict[str, Any]:
        sensor = to_config_dict(self.sensor) if self.sensor else None
        scenario = to_config_dict(self.scenario) if self.scenario else None

        config = {
            "sensor": sensor,
            "scenario": scenario,
            "mtf_wavelengths": self.mtf_wavelengths,
            "mtf_weights": self.mtf_weights,
            "interp": self.interp,
        }

        return config

    @classmethod
    def is_usable(cls) -> bool:
        # Requires opencv to be installed
        return cv2_available
