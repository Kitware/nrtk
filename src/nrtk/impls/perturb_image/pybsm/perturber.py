from __future__ import annotations

import copy
from importlib.util import find_spec
from typing import Any, Dict, Optional, Type, TypeVar

import numpy as np
from pybsm.simulation import RefImage, simulate_image
from smqtk_core.configuration import (
    from_config_dict,
    make_default_config,
    to_config_dict,
)

from nrtk.impls.perturb_image.pybsm.scenario import PybsmScenario
from nrtk.impls.perturb_image.pybsm.sensor import PybsmSensor
from nrtk.interfaces.perturb_image import PerturbImage

C = TypeVar("C", bound="PybsmPerturber")

DEFAULT_REFLECTANCE_RANGE = np.array([0.05, 0.5])  # It is bad standards to call np.array within argument defaults


class PybsmPerturber(PerturbImage):
    def __init__(
        self,
        sensor: PybsmSensor,
        scenario: PybsmScenario,
        reflectance_range: np.ndarray = DEFAULT_REFLECTANCE_RANGE,
        **kwargs: Any,
    ) -> None:
        """Initializes the PybsmPerturber.

        :param sensor: pyBSM sensor object.
        :param scenario: pyBSM scenario object.
        :param reflectance_range: Array of reflectances that correspond to pixel values.

        :raises: ValueError if reflectance_range length != 2
        :raises: ValueError if reflectance_range not strictly ascending
        """
        if not self.is_usable():
            raise ImportError("OpenCV not found. Please install 'nrtk[graphics]' or 'nrtk[headless]'.")
        self.sensor = copy.deepcopy(sensor)
        self.scenario = copy.deepcopy(scenario)

        for k in kwargs:
            if hasattr(self.sensor, k):
                setattr(self.sensor, k, kwargs[k])
            elif hasattr(self.scenario, k):
                setattr(self.scenario, k, kwargs[k])

        if reflectance_range.shape[0] != 2:
            raise ValueError(f"Reflectance range array must have length of 2, got {reflectance_range.shape[0]}")
        if reflectance_range[0] >= reflectance_range[1]:
            raise ValueError(f"Reflectance range array values must be strictly ascending, got {reflectance_range}")
        self.reflectance_range = reflectance_range

        # this is key:value record of the thetas use for perturbing
        self.thetas = copy.deepcopy(kwargs)

    @property
    def params(self) -> Dict:
        return self.thetas

    def perturb(self, image: np.ndarray, additional_params: Optional[Dict[str, Any]] = None) -> np.ndarray:
        """:raises: ValueError if 'img_gsd' not present in additional_params"""
        if additional_params is None:  # Cannot have mutable data structure in argument default
            additional_params = dict()
        if "img_gsd" not in additional_params:
            raise ValueError("'img_gsd' must be present in image metadata for this perturber")

        ref_img = RefImage(
            image,
            additional_params["img_gsd"],
            np.array([image.min(), image.max()]),
            self.reflectance_range,
        )

        perturbed = simulate_image(ref_img, self.sensor(), self.scenario())[-1]

        min_perturbed_val = perturbed.min()
        den = perturbed.max() - min_perturbed_val
        perturbed -= min_perturbed_val
        perturbed /= den
        perturbed *= 255

        return perturbed.astype(np.uint8)

    def __call__(self, image: np.ndarray, additional_params: Optional[Dict[str, Any]] = None) -> np.ndarray:
        """Alias for :meth:`.NIIRS.apply`."""
        if additional_params is None:
            additional_params = dict()
        return self.perturb(image, additional_params)

    def __str__(self) -> str:
        return self.sensor.name + " " + self.scenario.name

    def __repr__(self) -> str:
        return self.sensor.name + " " + self.scenario.name

    @classmethod
    def get_default_config(cls) -> Dict[str, Any]:
        cfg = super().get_default_config()
        cfg["sensor"] = make_default_config([PybsmSensor])
        cfg["scenario"] = make_default_config([PybsmScenario])
        cfg["reflectance_range"] = cfg["reflectance_range"].tolist()

        return cfg

    @classmethod
    def from_config(cls: Type[C], config_dict: Dict, merge_default: bool = True) -> C:
        config_dict = dict(config_dict)

        config_dict["sensor"] = from_config_dict(config_dict["sensor"], [PybsmSensor])
        config_dict["scenario"] = from_config_dict(config_dict["scenario"], [PybsmScenario])

        # Convert input data to expected constructor types
        config_dict["reflectance_range"] = np.array(config_dict["reflectance_range"])

        return super().from_config(config_dict, merge_default=merge_default)

    def get_config(self) -> Dict[str, Any]:
        config = {
            "sensor": to_config_dict(self.sensor),
            "scenario": to_config_dict(self.scenario),
            "reflectance_range": self.reflectance_range.tolist(),
        }

        return config

    @classmethod
    def is_usable(cls) -> bool:
        # Requires pyBSM which requires opencv to be installed
        return find_spec("cv2") is not None
