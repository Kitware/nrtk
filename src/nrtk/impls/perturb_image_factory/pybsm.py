from __future__ import annotations

from importlib.util import find_spec
from typing import Any, Dict, Iterable, Iterator, List, Sequence, Type, TypeVar

from smqtk_core.configuration import (
    from_config_dict,
    make_default_config,
    to_config_dict,
)

from nrtk.impls.perturb_image.pybsm.perturber import PybsmPerturber
from nrtk.impls.perturb_image.pybsm.scenario import PybsmScenario
from nrtk.impls.perturb_image.pybsm.sensor import PybsmSensor
from nrtk.interfaces.perturb_image import PerturbImage
from nrtk.interfaces.perturb_image_factory import PerturbImageFactory

C = TypeVar("C", bound="_PybsmPerturbImageFactory")


class _PybsmPerturbImageFactory(PerturbImageFactory):
    @staticmethod
    def _build_set_list(layer: int, top: Sequence[int]) -> Sequence[List[int]]:
        if layer == len(top) - 1:
            return [[i] for i in range(top[layer])]
        out = []
        for i in range(top[layer]):
            rec = _PybsmPerturbImageFactory._build_set_list(layer + 1, top)
            for e in rec:
                out.append([i] + e)
        return out

    def __init__(
        self,
        sensor: PybsmSensor,
        scenario: PybsmScenario,
        theta_keys: Iterable[str],
        thetas: Sequence[Any],
    ) -> None:
        """Initializes the PybsmPerturbImageFactory.

        :param sensor: pyBSM sensor object.
        :param scenario: pyBSM scenario object.
        :param theta_keys: Perturber parameter(s) to vary between instances.
        :param theta_keys: Perturber parameter(s) values to vary between instances.
        """
        if not self.is_usable():
            raise ImportError("OpenCV not found. Please install 'nrtk[graphics]' or 'nrtk[headless]'.")
        self.sensor = sensor
        self.scenario = scenario
        self.theta_keys = theta_keys
        self._thetas = thetas

        top = [len(entry) for entry in self.thetas]
        self.sets = _PybsmPerturbImageFactory._build_set_list(0, top)

    def __len__(self) -> int:
        return len(self.sets)

    def __iter__(self) -> Iterator[PerturbImage]:
        self.n = 0
        return self

    def __next__(self) -> PerturbImage:
        if self.n < len(self.sets):
            kwargs = {k: self.thetas[i][self.sets[self.n][i]] for i, k in enumerate(self.theta_keys)}
            func = PybsmPerturber(self.sensor, self.scenario, **kwargs)
            self.n += 1
            return func
        else:
            raise StopIteration

    def __getitem__(self, idx: int) -> PerturbImage:
        if idx >= len(self.sets):
            raise IndexError("Index out of range")
        kwargs = {k: self.thetas[i][self.sets[idx][i]] for i, k in enumerate(self.theta_keys)}
        func = PybsmPerturber(self.sensor, self.scenario, **kwargs)
        return func

    @property
    def thetas(self) -> Sequence[Sequence[Any]]:
        return self._thetas

    @property
    def theta_key(self) -> str:
        return "params"

    @classmethod
    def get_default_config(cls) -> Dict[str, Any]:
        cfg = super().get_default_config()

        # Remove perturber key if it exists
        cfg.pop("perturber", None)

        cfg["sensor"] = make_default_config([PybsmSensor])
        cfg["scenario"] = make_default_config([PybsmScenario])

        return cfg

    @classmethod
    def from_config(
        cls: Type[C],
        config_dict: Dict,
        merge_default: bool = True,
    ) -> C:
        config_dict = dict(config_dict)

        config_dict["sensor"] = from_config_dict(config_dict["sensor"], [PybsmSensor])
        config_dict["scenario"] = from_config_dict(config_dict["scenario"], [PybsmScenario])

        return super().from_config(config_dict, merge_default=merge_default)

    def get_config(self) -> Dict[str, Any]:
        return {
            "theta_keys": self.theta_keys,
            "sensor": to_config_dict(self.sensor),
            "scenario": to_config_dict(self.scenario),
            "thetas": self.thetas,
        }

    @classmethod
    def is_usable(cls) -> bool:
        # Requires pyBSM which requires opencv to be installed
        return find_spec("cv2") is not None


class CustomPybsmPerturbImageFactory(_PybsmPerturbImageFactory):
    def __init__(
        self,
        sensor: PybsmSensor,
        scenario: PybsmScenario,
        theta_keys: Sequence[str],
        thetas: Sequence[Any],
    ):
        super().__init__(sensor, scenario, theta_keys, thetas)
