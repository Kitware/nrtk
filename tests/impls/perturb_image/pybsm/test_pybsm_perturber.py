from contextlib import nullcontext as does_not_raise
from typing import Any, ContextManager, Dict

import numpy as np
import pytest
from PIL import Image
from smqtk_core.configuration import configuration_test_helper

from nrtk.impls.perturb_image.pybsm.perturber import PybsmPerturber

from ...test_pybsm_utils import create_sample_sensor_and_scenario
from ..test_perturber_utils import pybsm_perturber_assertions

INPUT_IMG_FILE = (
    "./examples/pybsm/data/M-41 Walker Bulldog (USA) width 319cm height 272cm.tiff"
)
EXPECTED_IMG_FILE = "./tests/impls/perturb_image/pybsm/data/Expected Output.tiff"


class TestPyBSMPerturber:
    def test_consistency(self) -> None:
        """Run on a dummy image to ensure output matches precomputed results."""
        image = np.array(Image.open(INPUT_IMG_FILE))
        expected = np.array(Image.open(EXPECTED_IMG_FILE))
        img_gsd = 3.19 / 160.0
        sensor, scenario = create_sample_sensor_and_scenario()
        # Test perturb interface directly
        inst = PybsmPerturber(sensor=sensor, scenario=scenario, ground_range=10000)
        pybsm_perturber_assertions(
            perturb=inst.perturb,
            image=image,
            expected=expected,
            additional_params={"img_gsd": img_gsd},
        )

        # Test callable
        pybsm_perturber_assertions(
            perturb=PybsmPerturber(
                sensor=sensor, scenario=scenario, ground_range=10000
            ),
            image=image,
            expected=expected,
            additional_params={"img_gsd": img_gsd},
        )

    @pytest.mark.parametrize(
        ("param_name", "param_value"),
        [
            ("ground_range", 10000),
            ("ground_range", 20000),
            ("ground_range", 30000),
            ("altitude", 10000),
            ("ihaze", 2),
        ],
    )
    def test_reproducibility(self, param_name: str, param_value: Any) -> None:
        """Ensure results are reproducible."""
        # Test perturb interface directly
        image = np.array(Image.open(INPUT_IMG_FILE))
        sensor, scenario = create_sample_sensor_and_scenario()
        inst = PybsmPerturber(
            sensor=sensor, scenario=scenario, **{param_name: param_value}
        )
        img_gsd = 3.19 / 160.0
        out_image = pybsm_perturber_assertions(
            perturb=inst.perturb,
            image=image,
            expected=None,
            additional_params={"img_gsd": img_gsd},
        )
        pybsm_perturber_assertions(
            perturb=inst.perturb,
            image=image,
            expected=out_image,
            additional_params={"img_gsd": img_gsd},
        )

    def test_configuration(self) -> None:
        """Test configuration stability."""
        sensor, scenario = create_sample_sensor_and_scenario()
        inst = PybsmPerturber(sensor=sensor, scenario=scenario)
        for i in configuration_test_helper(inst):
            assert i.sensor.name == sensor.name
            assert i.sensor.D == sensor.D
            assert i.sensor.f == sensor.f
            assert i.sensor.p_x == sensor.p_x
            assert np.array_equal(
                i.sensor.opt_trans_wavelengths, sensor.opt_trans_wavelengths
            )
            assert np.array_equal(
                i.sensor.optics_transmission, sensor.optics_transmission
            )
            assert i.sensor.eta == sensor.eta
            assert i.sensor.w_x == sensor.w_x
            assert i.sensor.w_y == sensor.w_y
            assert i.sensor.int_time == sensor.int_time
            assert i.sensor.n_tdi == sensor.n_tdi
            assert i.sensor.dark_current == sensor.dark_current
            assert i.sensor.read_noise == sensor.read_noise
            assert i.sensor.max_n == sensor.max_n
            assert i.sensor.bit_depth == sensor.bit_depth
            assert i.sensor.max_well_fill == sensor.max_well_fill
            assert i.sensor.s_x == sensor.s_x
            assert i.sensor.s_y == sensor.s_y
            assert i.sensor.da_x == sensor.da_x
            assert i.sensor.da_y == sensor.da_y
            assert np.array_equal(i.sensor.qe_wavelengths, sensor.qe_wavelengths)
            assert np.array_equal(i.sensor.qe, sensor.qe)

            assert i.scenario.name == scenario.name
            assert i.scenario.ihaze == scenario.ihaze
            assert i.scenario.altitude == scenario.altitude
            assert i.scenario.ground_range == scenario.ground_range
            assert i.scenario.aircraft_speed == scenario.aircraft_speed
            assert i.scenario.target_reflectance == scenario.target_reflectance
            assert i.scenario.target_temperature == scenario.target_temperature
            assert i.scenario.background_reflectance == scenario.background_reflectance
            assert i.scenario.background_temperature == scenario.background_temperature
            assert i.scenario.ha_wind_speed == scenario.ha_wind_speed
            assert i.scenario.cn2_at_1m == scenario.cn2_at_1m

            assert np.array_equal(i.reflectance_range, inst.reflectance_range)

    @pytest.mark.parametrize(
        ("reflectance_range", "expectation"),
        [
            (np.array([0.05, 0.5]), does_not_raise()),
            (np.array([0.01, 0.5]), does_not_raise()),
            (
                np.array([0.05]),
                pytest.raises(
                    ValueError, match=r"Reflectance range array must have length of 2"
                ),
            ),
            (
                np.array([0.5, 0.05]),
                pytest.raises(
                    ValueError,
                    match=r"Reflectance range array values must be strictly ascending",
                ),
            ),
        ],
    )
    def test_configuration_bounds(
        self, reflectance_range: np.ndarray, expectation: ContextManager
    ) -> None:
        """Test that an exception is properly raised (or not) based on argument value."""
        sensor, scenario = create_sample_sensor_and_scenario()
        with expectation:
            PybsmPerturber(
                sensor=sensor, scenario=scenario, reflectance_range=reflectance_range
            )

    @pytest.mark.parametrize(
        ("additional_params", "expectation"),
        [
            ({"img_gsd": 3.19 / 160.0}, does_not_raise()),
            (
                {},
                pytest.raises(
                    ValueError,
                    match=r"'img_gsd' must be present in image metadata for this perturber",
                ),
            ),
        ],
    )
    def test_additional_params(
        self, additional_params: Dict[str, Any], expectation: ContextManager
    ) -> None:
        """Test variations of additional params."""
        sensor, scenario = create_sample_sensor_and_scenario()
        perturber = PybsmPerturber(
            sensor=sensor, scenario=scenario, reflectance_range=np.array([0.05, 0.5])
        )
        image = np.array(Image.open(INPUT_IMG_FILE))
        with expectation:
            _ = perturber(image, additional_params)
