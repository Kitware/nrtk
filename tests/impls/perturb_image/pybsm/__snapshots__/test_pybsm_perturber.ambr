# serializer version: 1
# name: TestPyBSMPerturber.test_perturb_with_boxes[None]
  None
# ---
# name: TestPyBSMPerturber.test_perturb_with_boxes[boxes1]
  list([
    tuple(
      <smqtk_image_io.bbox.AxisAlignedBoundingBox min_vertex=[0. 0.] max_vertex=[20.1796875 38.71875  ]>,
      dict({
        'test': 0.0,
      }),
    ),
  ])
# ---
# name: TestPyBSMPerturber.test_perturb_with_boxes[boxes2]
  list([
    tuple(
      <smqtk_image_io.bbox.AxisAlignedBoundingBox min_vertex=[21.65625  36.421875] max_vertex=[72.84375 57.75   ]>,
      dict({
        'test': 0.7,
      }),
    ),
    tuple(
      <smqtk_image_io.bbox.AxisAlignedBoundingBox min_vertex=[37.8984375 18.2109375] max_vertex=[54.6328125 34.78125  ]>,
      dict({
        'test2': 1.0,
      }),
    ),
  ])
# ---
